### ISSUE TITLE
A GATK4.0 Spark bug about HaplotypeCallerSpark and How to fix

### ISSUE THREAD

**Issue by wangdy12** [broadinstitute/gatk#4231]
I hava try to run HaplotypeCallerSpark several times on a cluster. But the results were different with same inputs.  I found the results lost some varients in blocks. After debug, I have known the reason.

1. I found the results of `JavaRDD<Shard<GATKRead>> readShards`  at **HaplotypeCallerSpark.callVariantsWithHaplotypeCaller** were different.
2. `org.broadinstitute.hellbender.engine.spark.SparkSharder.joinOverlapping(JavaSparkContext , JavaRDD<L> locatables, Class<L> , SAMSequenceDictionary , List<I> intervals, int , FlatMapFunction2<Iterator<L>, Iterator<I>, T> f)`  wants to make shard and reads together. However **read is ordered but shards not**, the function `locatablesPerShard` requires shards must be ordered too.

 ## HOW TO FIX
change the code in `SparkSharder`
### First make shards  be ordered in each partition
old 
```
        JavaRDD<I> intervalsRdd = ctx.parallelize(indexedIntervals)
                .mapToPair(interval ->
                        new Tuple2<>(interval.getPartitionIndex(), interval.getLocatable()))
                .partitionBy(new KeyPartitioner(locatables.getNumPartitions())).values();
```

new
```
        JavaRDD<I> intervalsRdd = ctx.parallelize(indexedIntervals)
                .mapToPair(interval ->
                        new Tuple2<>(interval, interval.getLocatable()))
                .repartitionAndSortWithinPartitions(new KeyPartitioner(locatables.getNumPartitions()),
                        (Comparator<PartitionLocatable<I>> & Serializable) (o1, o2) -> {
                             return IntervalUtils.compareLocatables(o1, o2, sequenceDictionary);
                        }
                )
                .values();
```
### Then change inner  class `KeyPartitioner` a little
old
```
        @Override
        public int getPartition(Object key) {
            return (Integer) key;
        }
```
new
```
        @Override
        public int getPartition(Object key) {
            return ((PartitionLocatable) key).getPartitionIndex();
        }
```

**Comment by droazen** [comment]
@lbergelson Could you review this proposed fix?

**Comment by wangdy12** [comment]
First let's talk about the algorithm  in function `SparkSharder.shard`
Main inputs  : `JavaSparkContext ctx,  JavaRDD<L>  reads, List<ShardBoundary> intervals`
Output: `ShardBoundaryShard`
The reads have original partitions, but we want to make each interval and  its corresponding reads in the same partition

1.  Find out the first read for each partition, then compute out each partition extents with partitionIndex  `partitionReadExtents`
2. Compare the `partitionReadExtents` with `intervals`, some intervals overlap several partitions,  the infor is recorded in  `maxEndPartitionIndexes`
3.  Coalesce partitions, make each shard  correspond to one partition only
4. Put each interval to rigth partition whose  `partitionIndex` is the same as its  corresponding reads partition index
5. ZipPartitions: For each partition, traverse reads and intervals in order, finally output `ShardBoundaryShard`

Step 5 is done by `locatablesPerShard` function, which traverses reads and intervals iterators in order.
But step 4 only makes sure intervals all at right partitions,  the order in each partition can not be guaranteed.

So some intervals are not in order especially when we run code on a cluster.  Although the driver sends `indexedIntervals` in order, the order of intervals in each executor can not be guaranteed becasuse of the transfer delay on the network.

If you run the code in  local mode whose master URL is equal to `local[*]`, it difficult to find this bug because these is only one process, where delay is very small and can hardly influence the order.

But if on a cluster of multiple computers, the bug will appear, some intervals/shards are ignored, and the vcf results are diffierent each time.

**Comment by lelqz** [comment]
I got the the same issue。

**Comment by lelqz** [comment]
@wangdy12 .I get the same issue: when use HaplotypeCallerSpark on  a cluster, It lose  a lot of variable sites and the result jitter to the same input bam. and running on local mode, the result is good.
        But I fix the code according to your way. It does not work and get the same issue.

**Comment by wangdy12** [comment]
You'd better verify the consistency of the results step by step, then you can find which step's result become diffierent, finally fix it.
For example, you want to see if the `JavaRDD<Shard<GATKRead>> readShards`results are always same. You can add debug info in `callVariantsWithHaplotypeCaller` just like this:

```
......  final JavaRDD<Shard<GATKRead>> readShards = SparkSharder.shard ....
        //Debug
        String path = "./dubug/readShardsTest1";
        java.io.File debug = new File(path);
        if (!debug.getParentFile().exists())
            debug.getParentFile().mkdir();
        try (PrintStream printStream = new PrintStream(debug)) {
            printStream.println("List<ShardBoundary> shardBoundaries size : " + shardBoundaries.size());
            printStream.printf("NumPartitions : %d\n", readShards.getNumPartitions());

            List<ShardDebug> shardDebugs = readShards.mapToPair(shard -> new Tuple2<>(new ShardDebug(shard), null))
                    .sortByKey((Comparator<ShardDebug> & Serializable) (o1, o2) ->
                            IntervalUtils.compareLocatables(o1, o2, header.getSequenceDictionary())
                    ).keys().collect();
            printStream.printf("NumShard : %d\n", shardDebugs.size());
            for (ShardDebug shardDebug : shardDebugs) {
                printStream.println(shardDebug.toString());
            }
        }catch (Exception e){
            e.printStackTrace();
        }
```

```
    static class ShardDebug extends ShardBoundary{
        int size;

        public ShardDebug(Shard<GATKRead> shard) {
            super(shard.getInterval(),shard.getPaddedInterval());
            size = Iterators.size(shard.iterator());
        }

        @Override
        public String toString() {
            return this.getInterval().toString() + "\t" + size;
        }
    }
```
 
Run twice and compare the differences, do this step by step, you will find the bug. Gook Luck!

**Comment by lelqz** [comment]
@wangdy12 thanks.I have tested and your fix is effective. By Calling HaplotypeCallerSpark I can get the good results on an cluster. The GATK version is 4.0.2.1-2.

**Comment by droazen** [comment]
@jamesemery @jonn-smith @tomwhite You guys should have a look at this proposed fix as part of your 4Q work on `HaplotypeCallerSpark`

**Comment by tomwhite** [comment]
This was fixed in #5475 and released in 4.1.0.0. HaplotypeCallerSpark can be run in "strict" mode (with the `--strict` flag) to closely match the walker version. Even when not run using strict mode the result is much better than it used to be (closer to the walker version, and shouldn't jitter).
